Capsule improve ux

krystian-hebel opened the pull request:
Position and size of progress bar produced by DisplayUpdateProgressLibGraphics
depends on the size of boot logo. As the logo can be customized by users,
a dummy logo with determinable dimensions, depending on the size of current
GOP mode, is created to make progress bar appear always in the same place.

krystian-hebel committed:
DasharoPayloadPkg/CapsuleSplashDxe: add driver for consistent progress bar

Position and size of progress bar produced by
DisplayUpdateProgressLibGraphics depends on the size of boot logo.
As the logo can be customized by users, a dummy logo with determinable
dimensions, depending on the size of current GOP mode, is created to
make progress bar appear always in the same place.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CapsuleSplashDxe/SSFN: add scalable-font2 renderer

Project page: https://gitlab.com/bztsrc/scalable-font2

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
CapsuleSplashDxe/CapsuleSplashDxe.c: print some warnings

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

SergiiDmytruk commented on DasharoPayloadPkg/CapsuleSplashDxe/CapsuleSplashDxe.c:
```diff
@@ -171,16 +288,58 @@ CapsuleSplashEntry (
+  Status = RenderTextCenteredAt("Don't turn off your platform!",
+                                ModeInfo->HorizontalResolution / 2,
+                                ModeInfo->VerticalResolution * 90 / 100);
+  if (EFI_ERROR (Status)) {
```
Unless you intended to add some logging, this can be dropped.

krystian-hebel replied:
Removed both checks in https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371L333, https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371L340

SergiiDmytruk commented on DasharoPayloadPkg/CapsuleSplashDxe/CapsuleSplashDxe.c:
```diff
@@ -171,16 +288,58 @@ CapsuleSplashEntry (
+  //
+  // Clear the background. We don't know what logo was shown there, if any.
+  //
+  SetMem ((VOID *)GraphicsOutput->Mode->FrameBufferBase,
```
I forget to use it, but there is actually `ZeroMem()`.

krystian-hebel replied:
Right, that was the original code but for debugging I set it to grey and forgot to revert to `ZeroMem`.

https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371R327

SergiiDmytruk reviewed (commented):
Looks good.  Will need to test it on hardware as well and also want to check what happens on error in the new DXE.

SergiiDmytruk commented on DasharoPayloadPkg/CapsuleSplashDxe/CapsuleSplashDxe.c:
```diff
@@ -0,0 +1,345 @@
+
+  Status = RenderTextCenteredAt("Don't turn off your platform!",
+                                ModeInfo->HorizontalResolution / 2,
+                                ModeInfo->VerticalResolution * 90 / 100);
```
I think these percentages could be adjusted (this one to 88 and above one to 37) to align things better and instead of

![](https://private-user-images.githubusercontent.com/40995177/370257685-67a1be33-841d-4db0-8b5c-db390911a3f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyMTMzMjcsIm5iZiI6MTcyNzIxMzAyNywicGF0aCI6Ii80MDk5NTE3Ny8zNzAyNTc2ODUtNjdhMWJlMzMtODQxZC00ZGIwLThiNWMtZGIzOTA5MTFhM2YyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDIxMjM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxMDcyOTBjODE0NmRjYzEzZTQ0ZmM2ODczZDdmNjVhN2MzZGMzMjFlZjlhMmQ5MTcwODQ0MTdiOTdkZTQ3Y2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9QLfc7peQR-tSAsgb8hmu6hOJkcDyNlUZlU2kGa4NTE) get 
![ux](https://github.com/user-attachments/assets/c70c868c-e7db-4e51-863a-e9768259127f)

krystian-hebel replied:
Moved in https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371R336 and https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371R340.

I planned to add a glyph for Dasharo logo and print it in the middle, but it requires simplified SVG (just lines and Bezier curves, no arcs etc.) While working on it I discovered that printing the text would return the error on too big font size, and while it would require enormous resolution to trigger for existing lines, I added the code to clamp it anyway: https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371R169-R172

SergiiDmytruk replied:
Out of curiosity, why did you try to draw it as a glyph instead of a bitmap as the logo is drawn?

krystian-hebel replied:
Just for proper scaling. We will soon have a platform with 2880x1800 display, the usual logo would be too small for that.

SergiiDmytruk reviewed (commented):
If an error occurs, the update process seems to be completely aborted.  We might want to return `EFI_SUCCESS` on errors as well if that turns out to be a problem.

krystian-hebel committed:
CapsuleSplashDxe/CapsuleSplashDxe.c: move text on the screen, ignore errors

Reasoning for ignoring errors is that this driver gives only
informational output, and failure to do so shouldn't abort
the update process.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
